Improve and document text utility classes

lindsaym-fa opened the pull request:
Closes #1583 

- [x] Redefine `.wa-body`, `.wa-heading`, and `.wa-caption` classes so that the appended size corresponds exactly to the applied font size (e.g. `.wa-body-s`, `.wa-heading-s`, and `.wa-caption-s` are all styled with `font-size: var(--wa-font-size-s);`)
- [x] Add `.wa-body-*`, `.wa-heading-*`, and `.wa-caption-*` classes for all font size tokens
- [x] Find and replace all prior uses of `.wa-body-*`, `.wa-heading-*`, and `.wa-caption-*` classes with the adjusted sizes to maintain the intended font size (requires a Pro PR)
- [x] Add `.wa-longform-*` classes for all font size tokens.
- [x] Add single-purpose `.wa-font-size-*` classes for all font size tokens
- [x] Add single-purpose `.wa-color-text-*` classes for all text color tokens
- [x] Add multi-purpose `.wa-form-control-*` classes to emulate form control labels, values, placeholders, and hints
- [x] Document text utilities

---

To test the related changes across all repos, pull:
- Free: `lm/improve-text-utilities` (this!)
- App: `lm/update-text-utilities-app`(https://github.com/shoelace-style/webawesome-app/pull/168)
- Pro: `lm/update-text-utilities-pro` (https://github.com/shoelace-style/webawesome-pro/pull/67)

lindsaym-fa committed:
initial take on text utility improvements

lindsaym-fa committed:
Merge branch 'next' into lm/improve-text-utilities

lindsaym-fa committed:
re-introduce `.wa-caption-*`

lindsaym-fa committed:
Merge branch 'next' into lm/improve-text-utilities

lindsaym-fa committed:
reorg stylesheet

lindsaym-fa committed:
remove font weight classes

talbs reviewed (commented):
@lindsaym-fa, thanks for the good work!

This direction makes sense to me. **No reservations!**

Having the class names' suffixes match the literal t-shirt sizing feels more straight-forward to me. I'd be more confident using these and knowing how they'd render.

> _Also, are there any other text utilities you feel would be beneficial, given your experience styling patterns and the app?_

Hmm, here are a few places I thought we could expand...

- **Extending the size scale:** I have more than twice rolled my own larger `font-sizes` based on our scales multiplier (> `4xl`). I think I may have also added a step on the smaller side of the scale as well (to get down to ~10px). Unrelated, I think adding sizes on the larger side of our `wa-space` scale would be good as well.
- **Font Weight Utilities:** I noticed you removed these utilities in https://github.com/shoelace-style/webawesome/pull/1602/commits/cdd1bcebcf02790195f893f3f644d68cd2f1c797. They seemed useful and inline with other levels of text-based utilities we were including.
- **Code Font Styling:** Our `<code>` and monospace-based font styling renders a bit larger than our other text when rendered side-by-side. I've written some custom CSS here and there to "reset" that so that it is vertically and visually balanced.
- **Volume-level Utilities:** I see you added `wa-color-text-quiet` and `wa-color-text-normal`. Is there a place/need for a `.wa-color-text-loud` to mirror the general levels of volume we have documented? Or is `loud` superfluous given that `normal` already uses the farther side of our color scales? 
- **More Text-Wrangling Utilities:** I started [a few simple utilities](https://github.com/shoelace-style/webawesome/blob/next/packages/webawesome/docs/assets/styles/utils.css#L38) around text management in our utils.css file. I could see `word-break` and `text-wrap` shortcuts being handy.

---

Hope that helps. Shout if you wanna chat anything out in real time.

talbs commented on packages/webawesome/src/styles/utilities/text.css:
```diff
@@ -1,106 +1,98 @@
-  .wa-heading-2xl,
-  .wa-heading-3xl {
+  /* #region General ~~~~~ */
+  [class*='wa-body'] {
```
So fresh so clean clean! 🧽 🫧 ✨

talbs commented on packages/webawesome/src/styles/utilities/text.css:
```diff
@@ -1,106 +1,98 @@
+    line-height: var(--wa-line-height-normal);
+  }
+
+  [class*='wa-heading'] {
```
Dope clean up 🧹 🧼

talbs commented on packages/webawesome/src/styles/utilities/text.css:
```diff
@@ -1,106 +1,98 @@
-
-  .wa-heading-2xl {
-    font-size: var(--wa-font-size-3xl);
+  [class*='wa-caption'] {
```
💪

talbs commented on packages/webawesome/src/styles/utilities/text.css:
```diff
@@ -1,106 +1,98 @@
 
-  .wa-heading-3xl {
-    font-size: var(--wa-font-size-4xl);
+  .wa-font-size-2xs,
```
🧠

talbs commented on packages/webawesome/src/styles/utilities/text.css:
```diff
@@ -126,4 +118,26 @@
   }
+  /* #endregion */
+
+  /* #region Form Controls */
```
```suggestion
  /* #region Form Controls ~~~~~~~ */
```

You were missing them squigglies here. Need 'em? Also, what do they do - just easy find in file business?

lindsaym-fa replied:
I use 'em to align the right edge of `#region` headings so that they have a clean edge when regions are collapsed. So, I'll add just enough squigglies to each line to match the length of the longest one, like so:
<img width="237" height="129" alt="Screenshot 2025-10-15 at 1 15 38 PM" src="https://github.com/user-attachments/assets/61b895ae-c3e9-45e3-8d6c-7730642ddeb5" />

The same convention is at play in `native.css`.

We could go no-squiggly if it's more burdensome/confusing than its worth, though. Thoughts?

talbs replied:
> _We could go no-squiggly if it's more burdensome/confusing than its worth, though. Thoughts?_

Now that I've learned of the squigglies, no-squiggly is a no-go for me.  I'll sync up the other regions in other CSS files if I see any out of whack in my travels!

talbs commented:
@lindsaym-fa, just a reminder that a "Find and replace all prior uses..." will span the entire turducken. Shout if you want help with that.

lindsaym-fa commented:
Thanks, @talbs!

> **Extending the size scale:** I have more than twice rolled my own larger `font-sizes` based on our scales multiplier (> `4xl`). I think I may have also added a step on the smaller side of the scale as well (to get down to ~10px). Unrelated, I think adding sizes on the larger side of our `wa-space` scale would be good as well.

Heard! We'll tackle this in #1606 to keep the scope of this work a bit slimmer. (Especially since those additions won't be breaking changes.)

> **Font Weight Utilities:** I noticed you removed these utilities in https://github.com/shoelace-style/webawesome/pull/1602/commits/cdd1bcebcf02790195f893f3f644d68cd2f1c797. They seemed useful and inline with other levels of text-based utilities we were including.

Sharp eye. I removed these because I was projecting usefulness, but your comment is giving _proven_ usefulness. I'll re-add these. 😄

> **Code Font Styling:** Our `<code>` and monospace-based font styling renders a bit larger than our other text when rendered side-by-side. I've written some custom CSS here and there to "reset" that so that it is vertically and visually balanced.

As discussed privately, this likely warrants an adjustment to our native styles so that we're rendering code elements in an even smaller relative font size than we already do. I'll investigate this separately.

> **Volume-level Utilities:** I see you added `wa-color-text-quiet` and `wa-color-text-normal`. Is there a place/need for a `.wa-color-text-loud` to mirror the general levels of volume we have documented? Or is `loud` superfluous given that `normal` already uses the farther side of our color scales? 

Readability of general text content warrants a tint at the far end of our color scales, true. There are other, non-color approaches to make text louder — such as **bold** or larger font sizes — that extend beyond our available `--wa-color-text-*` tokens. Text colors and fill/stroke colors are slightly different scales, so I'd say we can go without `.wa-color-text-loud`.

> **More Text-Wrangling Utilities:** I started [a few simple utilities](https://github.com/shoelace-style/webawesome/blob/next/packages/webawesome/docs/assets/styles/utils.css#L38) around text management in our utils.css file. I could see `word-break` and `text-wrap` shortcuts being handy.

Agreed! I'll hold off on those line length utilities for a time when we can add first-class design tokens for those, but `word-break` and `text-wrap` are simple and handy.

lindsaym-fa committed:
gimme the font weights

lindsaym-fa committed:
add `.wa-text-truncate`, `.wa-list-plain`

talbs commented:
@lindsaym-fa, all sounds good to me in [your previous comment](https://github.com/shoelace-style/webawesome/pull/1602#issuecomment-3407681858). `<wa-garth>Party on, Wayne!</wa-garth>`

Let me know when this branch (or others) are ready for a review if you need more 👀 .

lindsaym-fa committed:
Merge branch 'next' into lm/improve-text-utilities

lindsaym-fa committed:
add longform, color-text-link, and placeholder classes

lindsaym-fa committed:
document utilities

lindsaym-fa committed:
document missing font weight tokens

lindsaym-fa committed:
use proper longform font weight

lindsaym-fa committed:
find and replace outdated text utilities

lindsaym-fa committed:
use new font size utilities

lindsaym-fa committed:
missed a spot!

talbs reviewed (approved):
Docs and the find and replace efforts look good - nice work!

lindsaym-fa committed:
add changelog
